### PR TITLE
[StructuralMechanicsApplication] Adding slave solution reconstruction to eigensolver strategy

### PR DESCRIPTION
**Pull request by emiroglu**
**Description**
This PR adds reconstructing solution for slave DOFs to the Eigensolver Strategy for master-slave constraints, specifically for Block B&S. Without this addition, the slave DOF solution is not written to nodal `EIGENVECTOR_MATRIX` variable and the responsibility is shifted to the `postprocess_eigenvalues_process`.

**Changelog**
- Added ReconstructSolution function to `eigensolver_strategy.hpp`
- Deleted solution reconstruction from `postprocess_eigenvalues_process.cpp`

**Remark**
A much cleaner solution would be to add a function to BuilderAndSolver base class that recovers the slave DOF solution, which can then be called from the strategy directly. This is not done here, since the master-slave constraints don't work with all other B&S consistently yet.

### CODE REVIEW

**Comment by emiroglu** [comment]
In case of MS constraints the eigenvector solution in the `EigensystemSolver` is normalized without taking the slave solution into account. I think this might cause a problem in the system reduction -> 

Then, does it make sense to (ortho)normalize in strategy or solver?

What do you think @armingeiser @oberbichler ?

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
                    const auto& r_master_dofs_vector = it_const->GetMasterDofsVector();
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
                    const auto& r_slave_dofs_vector = it_const->GetSlaveDofsVector();
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
                const auto& r_slave_dofs_vector = it_const->GetSlaveDofsVector();
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
                for (IndexType i = 0; i < r_slave_dofs_vector.size(); ++i)
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
                    rEigenvectors(i_eigenvalue, r_slave_dofs_vector[i]->EquationId()) = 0.0;
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
                    Vector master_dofs_values(r_master_dofs_vector.size());
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
                    for (IndexType i = 0; i < r_master_dofs_vector.size(); ++i)
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
                        master_dofs_values[i] = rEigenvectors(i_eigenvalue, r_master_dofs_vector[i]->EquationId());
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
                        rEigenvectors(i_eigenvalue, r_slave_dofs_vector[i]->EquationId()) += aux;
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
                auto it_const = r_master_slave_constraints.begin() + i_const;
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
        const auto& r_master_slave_constraints = rModelPart.MasterSlaveConstraints();
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
        const std::size_t number_of_constraints = rModelPart.NumberOfMasterSlaveConstraints();
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
        const std::size_t number_of_eigenvalues = rEigenvectors.size1();
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
                auto it_const = r_master_slave_constraints.begin() + i_const;
```

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Great job, can you add a test?

**Comment by emiroglu** [comment]
@loumalouomega extended the existing test

**Comment by armingeiser** [comment]
> In case of MS constraints the eigenvector solution in the `EigensystemSolver` is normalized without taking the slave solution into account. I think this might cause a problem in the system reduction -> 
> 
> Then, does it make sense to (ortho)normalize in strategy or solver?
> 
> What do you think @armingeiser @oberbichler ?

Not sure, but i would assume that the normalization of the system that is used to solve the eigenvalues/vectors contains all the information from the constraints. What i mean is, even though the eigenvectors do not have entries for the slave dofs, the corresponding masses are considered at the master dofs in the mass matrix.

Also, how could this be done otherwise? I guess in the strategy after reconstructing the slave dofs, but then one would need to assemble a new mass matrix without taking the constraints into account?

**Review by loumalouomega — DISMISSED** [review]

**Comment by loumalouomega** [comment]
Ok from my side, but wait for an additional approval

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
     * @brief Recover the solution related to the slave dofs in case of master-slave constraints.
     * @details This function is an adaptation of the implementation in ConstraintUtilities,
```

**Comment by emiroglu** [comment]
@loumalouomega thanks. waiting for @philbucher

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
        if (master_slave_constraints_defined) {
            this->ReconstructSolution(Eigenvectors);
        }
```

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]: can you use the parallel utilities please?

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]: same

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]: this should be declared outside the loop

**Review by philbucher — COMMENTED** [review]
@emiroglu I am ok with the PR, but I have two requests:
- please make a PR to the Core B&S for this functionality. Then we can have a cleaner solution (Independent of this PR)
- Please use the parallelUtilites for the loops you add, as I pointed out. You should be able to use a simple `block_for_each` in the first case and one with Thread Local Storage (see #8195)

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (emiroglu) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]: added atomic update here too

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (emiroglu) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]: done

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (emiroglu) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]: done

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (emiroglu) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]: done

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (emiroglu) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]: done

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]: please use [this](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/atomic_utilities.h#L39) instead

Comment (emiroglu) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]: done

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]: the slave dof can exist several time no?
In this case this would be a race condition
=> do the same as [here](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/constraints/linear_master_slave_constraint.cpp#L46) in this case

(AtomicMult will only be added in #8155)

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
    void ReconstructSlaveSolution(
```

I think this is a more applicable name

**Review by philbucher — COMMENTED** [review]
minor otherwise ready :+1:

**Review by philbucher — APPROVED** [review]
Good Job :+1: 

as I said please open a PR with how you think it should be integrated into the base-B&S

**Comment by emiroglu** [comment]
> Good Job 
> 
> as I said please open a PR with how you think it should be integrated into the base-B&S

Thanks. That's the next TODO!

**Pull request merged**
